zeeguu/core/nlp_pipeline/reduce_context.py

```diff
@@ -0,0 +1,77 @@
+import heapq
+import numpy as np
+from .spacy_wrapper import SpacyWrapper
+
+class ContextReducer():
+    @classmethod
+    def get_similar_sentences(cls, nlp_pipe:SpacyWrapper, sentence:str, article:str, max_length:int = 15):
+        sentences = nlp_pipe.get_sent_list(article)
+        short_sentences_in_article = [sent for sent in sentences if len(nlp_pipe.tokenize_sentence(sent)) <= max_length]
+        
+        context_doc = nlp_pipe.get_doc(sentence)
+        heap = []
+        for f_sent in short_sentences_in_article:
+            sent_doc = nlp_pipe.get_doc(f_sent)
+            heapq.heappush(heap, (context_doc.similarity(sent_doc), f_sent))
+        
+        most_similar_sent = heapq.nlargest(1, heap)[0][1]
+        top_10_results = heapq.nlargest(10, heap)
+
+        result_json = {
+            "top_1_sent": most_similar_sent,
+            "top_10_sents_w_sim": top_10_results
+        }
+
+        return result_json
+    
+    @classmethod
+    def reduce_context_for_bookmark(cls, nlp_pipe:SpacyWrapper, sentence:str, bookmark:str, max_length:int = 15):
```

Comment:
for this one we should do another walkthrough next week maybe; i'd like to understand it :)